packages/docusaurus-theme-common/src/utils/useTOCHighlight.ts

```diff
@@ -138,19 +141,30 @@ function useTOCHighlight(config: TOCHighlightConfig | undefined): void {
       linkActiveClassName,
       minHeadingLevel,
       maxHeadingLevel,
+      autoScrollTOC,
     } = config;
 
-    function updateLinkActiveClass(link: HTMLAnchorElement, active: boolean) {
-      if (active) {
-        if (lastActiveLinkRef.current && lastActiveLinkRef.current !== link) {
-          lastActiveLinkRef.current?.classList.remove(linkActiveClassName);
-        }
-        link.classList.add(linkActiveClassName);
-        lastActiveLinkRef.current = link;
-        link.scrollIntoView({block: 'nearest'});
-      } else {
-        link.classList.remove(linkActiveClassName);
+    // When the page is scrolling (either a user scroll or smooth CSS scroll),
+    // We need to defer the TOC scroll or the two concurrent scrolls would block
+    // each other.
+    function scheduleScroll(link: HTMLAnchorElement) {
+      if (!autoScrollTOC) {
+        return () => {};
       }
+      const handle = setTimeout(() => {
+        const linkRect = link.getBoundingClientRect();
+        const viewport = window.visualViewport;
+        // Only scroll if a vertical scroll is sufficient to bring the link into
+        // view. e.g. if the window is pinch-zoomed, we should not horizontally
+        // scroll
+        if (
+          linkRect.right <= viewport.pageLeft + viewport.width &&
+          linkRect.left >= viewport.pageLeft
+        ) {
+          link.scrollIntoView({block: 'nearest', behavior: 'smooth'});
+        }
+      }, 200);
```

Comment:
magic timeout values like that are a recipe for troubles 😅 I'd like to avoid it at all cost

What about waiting for the end of scroll events using debounce + trailing events?

Reply:
Well, even if you are debouncing, you have to debounce within a certain time range, right?